bundle/src/main/java/com/adobe/acs/commons/mcp/impl/processes/BrokenLinksReport.java

```diff
@@ -0,0 +1,211 @@
+            options = {"base=/content"})
+    private String sourcePath;
+
+    @FormField(name = "Property Regex",
+            description = "Regular expression to detect properties containing references to Sling resources",
+            required = false,
+            options = {"default=^/(etc|content)/.+"})
+    private String propertyRegex;
+
+    @FormField(name = "Exclude Properties",
+            description = "Comma-separated list of properties to ignore",
+            hint = "cq:template,cq:allowedTemplates,....",
+            required = false,
+            options = {"default=cq:allowedTemplates"})
+    private String excludeProperties;
+
+    @FormField(
+            name = "Deep check in html",
+            description = "If checked, links will be extracted from html field",
+            component = CheckboxComponent.class,
+            options = {"checked"}
+    )
+    private boolean deepCheck = false;
+
+    @FormField(name = "Fields containing html",
+            description = "Properties containing html to extract links",
+            required = false,
+            options = {"default=text"})
+    private String htmlFields;
+
+    transient private Set<String> excludeList;
+    transient private Set<String> deepCheckList;
+    transient private Pattern regex;
+
+    @Override
+    public void init() throws RepositoryException {
+        excludeList = Arrays.stream(excludeProperties.split(",")).map(String::trim).collect(Collectors.toSet());
+        deepCheckList = deepCheck ? Arrays.stream(htmlFields.split(",")).map(String::trim).collect(Collectors.toSet())
+            : new HashSet<>();
+        regex = Pattern.compile(propertyRegex);
+    }
+
+    transient private final GenericReport report = new GenericReport();
+
+    enum REPORT {
+        reference
+    }
+
+    transient private final Map<String, EnumMap<REPORT, Object>> reportData = new TreeMap<>();
+
+    @Override
+    public void buildProcess(ProcessInstance instance, ResourceResolver rr) throws LoginException, RepositoryException {
+        report.setName(instance.getName());
+        instance.defineAction("Collect Broken References", rr, this::buildReport);
+        instance.getInfo().setDescription(sourcePath);
+
+    }
+
+    @Override
+    public void storeReport(ProcessInstance instance, ResourceResolver rr) throws RepositoryException, PersistenceException {
+        GenericReport report = new GenericReport();
+        report.setRows(reportData, "Source", REPORT.class);
+        report.persist(rr, instance.getPath() + "/jcr:content/report");
+
+    }
+
+    public void buildReport(ActionManager manager) {
+        TreeFilteringResourceVisitor visitor = new TreeFilteringResourceVisitor();
+        visitor.setBreadthFirstMode();
+        visitor.setTraversalFilter(null);
+        visitor.setResourceVisitor((resource, depth) -> {
+            Map<String, List<String>> brokenRefs = collectBrokenReferences(resource, regex, excludeList, deepCheckList);

```

Comment:
This part should be wrapped by manager.deferredWithResolver as well, this is how you can divide the work and really speed things up in parallel.  The first part being deferred is mostly so that you benefit from error handling more than anything else.

Reply:
A good catch. With this block of code wrapped in manager.deferredWithResolver the number of completed tasks is proportional to the number of visited ValueMaps. Each visit is a unit of work and can run in a pooled thread. Previously all the processing was single-threaded.